### PR TITLE
Cosmos encryption AKV integration initial version

### PR DESCRIPTION
**Pull request by moderakh**
This PR follows the design of https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1660
to provide an `AzureKeyVaultCosmosEncryptor`.

This PR doesn't change anything in the azure-cosmos. All new changes are in azure-cosmos-encryption.

TODOs to be addressed: (after this PR)
- akv blocking calls should become non-blocking
- sample missing
- code style fix, java convention to name static constant members, etc.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/AzureKeyVaultKeyWrapProvider.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+package com.azure.cosmos.encryption;
+
+import com.azure.cosmos.implementation.apachecommons.lang.StringUtils;
+import com.azure.cosmos.implementation.encryption.EncryptionUtils;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicReference;
+
+// TODO: moderakh remove blocking calls
+public class AzureKeyVaultKeyWrapProvider implements EncryptionKeyWrapProvider {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/AzureKeyVaultKeyWrapProvider.java:13]: We should add some description to these class.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        //        catch (RequestFailedException ex)
+        //        {
+        //            throw new KeyVaultAccessException(
+        //                ex.Status,
+        //                ex.ErrorCode,
+        //                "WrapKeyAsync: Failed to Wrap the data encryption key.",
+        //                ex);
+        //        }
+        catch (Exception ex) {
+            ex.printStackTrace();
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java:145]: logger ?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        //        catch (RequestFailedException ex)
+        //        {
+        //            throw new KeyVaultAccessException(
+        //                ex.Status,
+        //                ex.ErrorCode,
+        //                "ValidatePurgeProtectionAndSoftDeleteSettingsAsync: Failed to fetch Key from Key Vault.",
+        //                ex);
+        //        }
+        catch (Exception ex) {
+            ex.printStackTrace();
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java:190]: logger ?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessException.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    public KeyVaultAccessException(int statusCode,
+                                   String keyVaultErrorCode,
+                                   String errorMessage,
+                                   Exception innerException) {
+        // TODO: CosmosException as super class?
+        //super(statusCode, keyVaultErrorCode, errorMessage, innerException)
+    }
+}
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessException.java:22]: new line at the end of file.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultKeyUriProperties.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            keyVaultUriProperties = new KeyVaultKeyUriProperties(keyUri);
+            keyVaultUriProperties.keyName = segments[1]; // "testKey1/"
+            keyVaultUriProperties.keyVersion = segments[2]; // "47d306aeaae74baab294672354603ca3"
+            //keyVaultUriProperties.keyVaultUri = new URI(keyVaultUriProperties.KeyUri.GetLeftPart(UriPartial.Scheme
+            // | UriPartial.Authority)); // https://testdemo1.vault.azure.net/
+            keyVaultUriProperties.keyVaultUri = new URI(keyUri.getScheme(), keyUri.getAuthority(), null, null, null); // https://testdemo1.vault.azure.net/
+
+            keyVaultUriPropertiesReference.set(keyVaultUriProperties);
+        } catch (URISyntaxException e) {
+            return false;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultKeyUriProperties.java:73]: Do we need some sort of debug / warn error level logging here ?

File: sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/KeyModelFactory.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import com.azure.security.keyvault.keys.models.KeyProperties;
+import com.azure.security.keyvault.keys.models.KeyType;
+import com.azure.security.keyvault.keys.models.KeyVaultKey;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+
+public class KeyModelFactory {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/KeyModelFactory.java:18]: Please add public description of the class.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/AzureKeyVaultKeyWrapProvider.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/AzureKeyVaultKeyWrapProvider.java:17]: thanks. addressed.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java (hunk 3 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java]: thanks. addressed.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessClient.java]: thanks addressed.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessException.java (hunk 4 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultAccessException.java]: thanks addressed.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultKeyUriProperties.java (hunk 5 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/KeyVaultKeyUriProperties.java:77]: thanks. addressed.

File: sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/KeyModelFactory.java (hunk 6 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/KeyModelFactory.java:21]: thanks addressed.

**Pull request merged**